qiskit/algorithms/evolvers/real/trotterization/trotter_ops_validator.py

```diff
@@ -0,0 +1,131 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Set of method for validating input to TrotterQrte algorithm."""
+
+import numbers
+from typing import Union
+
+from qiskit.circuit import Parameter, ParameterExpression
+from qiskit.opflow import (
+    SummedOp,
+    PauliOp,
+)
+
+
+def is_op_bound(operator: Union[SummedOp, PauliOp]) -> None:
+    """Checks if an operator provided has all parameters bound.
+
+    Args:
+        operator: Operator to be checked.
+
+    Raises:
+        ValueError: If an operator has unbound parameters.
+    """
+    if len(operator.parameters) > 0:
+        raise ValueError(
+            f"Did not manage to bind all parameters in the Hamiltonian, "
+            f"these parameters encountered: {operator.parameters}."
+        )
+
+
+def validate_hamiltonian_form(hamiltonian: Union[SummedOp, PauliOp]):
```

Comment:
Generally the application stacks produce a PauliSumOp and that is by default what is created if you sum Paulis nowadays. The more general SummedOp, that can contain any sort of opflow operator, is not so common.

Reply:
Occurrences of `SummedOp` in this algorithm stem from the fact that ultimately (not in this PR) we will also support evolutions of gradients of states. If I understand correctly, `PauliSumOp` cannot be parametrized (except for maybe the global phase). Whereas in `SummedOp`, every summand can have a parametrized coefficient which gives users the possibility of defining such a parametrized summed state in evolving gradients of that state.

Reply:
So the assumption that we established for this algorithm with Stefan is to either have a single `PauliOp` or `SummedOp` of `PauliOp`s. You can see that in the validation we check if there are `PauliOp`s only in a `SummedOp`. 
But I support your argument about `PauliSumOp` and that it is widely used in the codebase. So perhaps I would add support for evolving `PauliSumOp`s and in case of gradients (in the future) it will be detected as not supported for gradients only (because of the lack of parametrization).